src/metabase/lib/equality.cljc

```diff
@@ -104,106 +109,365 @@
+           result                  {}]
+      (cond
+        (or (empty? current-needles)
+            (empty? current-haystack)
+            (nil? ref-fn))            result
+        (nil? ref-fn)                 (recur current-needles current-haystack more-ref-fns result)
+        :else
+        (let [haystack-ref->col  (m/index-by ref-fn current-haystack)
+              matches            (-> (find-closest-matches-for-refs current-needles (keys haystack-ref->col) opts)
+                                     ;; This returns a map {haystack-ref index-in-current-needles}.
+                                     ;; We want to adjust both keys and vals to
+                                     ;; {haystack-metadata index-in-original-needles}
+                                     (update-vals #(->> % (nth current-needles) meta ::index))
+                                     (update-keys haystack-ref->col))
+              matched-needles    (set (vals matches))]
+          (recur (remove #(matched-needles (::index (meta %))) current-needles)
+                 ;; Replace the matched haystacks with nil so they can't match again.
+                 (for [metadata current-haystack]
+                   (when-not (contains? matches metadata)
+                     metadata))
+                 more-ref-fns
+                 ;; TODO: Merging in this direction gives priority to the earliest match.
+                 ;; Perhaps any collision should be an error?
+                 (merge matches result)))))))
+
+(mu/defn closest-matches-in-metadata :- [:map-of
+                                         lib.metadata/ColumnMetadata
+                                         ::lib.schema.common/int-greater-than-or-equal-to-zero]
+  "Like [[find-closest-matches-for-refs]], but finds the closest match for each element of `needles` from a haystack of
+  Column `metadatas` rather than a haystack of refs. This allows us to do more sophisticated fuzzy matching than
+  [[find-closest-matches-for-refs]], because column metadatas inherently have more information than they do after they
+  are converted to refs.
+
+  Returns a map `{column-metadata index-in-needles}` for each successful match. If nothing is found that matches a
+  `needle`, no entry for it appears in the map.
+
+  If you only have a single `needle` ref to look up, use [[closest-matching-metadata]]."
+  ([needles haystack-metadatas]
+   (closest-matches-in-metadata needles haystack-metadatas {}))
+
+  ([needles haystack-metadatas opts]
+   (when (seq haystack-metadatas)
+     ;; The matching is done in two passes: first as plain refs, second by forcing ID refs.
+     ;; Each of these returns a map of `{^{::index i} haystack-ref needle-index}`. Those haystack values which match
+     ;; are removed between stages.
+     (cascading-matches
+       needles haystack-metadatas opts
+       [;; The first pass attempts to match the refs using the regular [[find-closest-matches-for-refs]].
+        lib.ref/ref
```

Comment:
I still have the feeling that this can result in a worse match than optimal by ignoring a good match by ID of a column from a source card (generating a ref by name) in favor of a worse match from a table (generating a ref by ID).